python/iceberg/parquet/parquet_reader.py

```diff
@@ -0,0 +1,240 @@
+     # To-Do: update to support reading map fields
+     # TypeID.MAP: lambda field: (,),
+     TypeID.STRING: lambda field: (pa.string(), ""),
+     TypeID.STRUCT: lambda field: (pa.struct([(nested_field.name,
+                                               DTYPE_MAP[nested_field.type.type_id](nested_field.type)[0])
+                                              for nested_field in field.type.fields]), {}),
+     TypeID.TIMESTAMP: lambda field: (pa.timestamp("us"), datetime.now()),
+     # not used in SPARK, so not implementing for now
+     # TypeID.TIME: pa.time64(None)
+     }
+
+
+class ParquetReader(object):
+
+    def __init__(self, input: InputFile, expected_schema: Schema, options, filter_expr: Expression,
+                 case_sensitive: bool, start: int = None, end: int = None):
+        self._stats: Dict[str, int] = dict()
+
+        self._input = input
+        self._input_fo = input.new_fo()
+
+        self._arrow_file = pq.ParquetFile(self._input_fo)
+        self._file_schema = convert_parquet_to_iceberg(self._arrow_file)
+        self._expected_schema = expected_schema
+        self._file_to_expected_name_map = ParquetReader.get_field_map(self._file_schema,
+                                                                      self._expected_schema)
+        self._options = options
+        self._filter = get_dataset_filter(filter_expr, ParquetReader.get_reverse_field_map(self._file_schema,
+                                                                                           self._expected_schema))
+
+        self._case_sensitive = case_sensitive
+        if start is not None or end is not None:
+            raise NotImplementedError("Partial file reads are not yet supported")
+            # self.start = start
+            # self.end = end
+
+        self.materialized_table = False
+        self.curr_iterator = None
+        self._table = None
+        self._df = None
+
+        _logger.debug("Reader initialized for %s" % self._input.path)
+
+    @property
+    def stats(self) -> dict:
+        return dict(self._stats)
+
+    def to_pandas(self) -> Union[pd.Series, pd.DataFrame]:
+        if not self.materialized_table:
+            self._read_data()
+            with profile("to_pandas", self._stats):
+                if self._table is not None:
+                    self._df = self._table.to_pandas(use_threads=True)
+                else:
+                    self._df = None
+
+        return self._df
+
+    def to_arrow_table(self) -> pa.Table:
+        if not self.materialized_table:
+            self._read_data()
+
+        return self._table
+
+    def _read_data(self) -> None:
+        _logger.debug("Starting data read")
+
+        # only scan the columns projected and in our file
+        cols_to_read = prune_columns(self._file_schema, self._expected_schema)
+
+        with profile("read data", self._stats):
+            arrow_dataset = ds.FileSystemDataset.from_paths([self._input.location()],
```

Comment:
Its a shame that we can't do this for a batch of files. Seems like a common pattern that pyarrow should be able to handle. Have you asked on user@arrow.apache.org if there is a plan to support this?

Do you have a sense of the performance hit of reading files individually as opposed to in one batch? I am worried it could be significant. 

Finally, what is the path for reading/iterating multi-file iceberg tables? Do we read 1 parrquet file at a time or all at once or all at once as a pandas df? One of the key features of Datasets (for me) was being able to control memory when doing large reads.

Reply:
I agree this is the biggest problem with this implementation. I can start a discussion on the arrow dev list and see if there are any ideas here.

Reply:
Here is the thread on the user list:
https://lists.apache.org/thread.html/r7d82d38d1e9a1b9bc99d64a0779de8886cf0bf7216d64d576a6c18e9%40%3Cuser.arrow.apache.org%3E

Reply:
currently, the way that I do it is to downloaded the needed files in parallel using an internal lib, then do a single threaded read of the local files into an arrow table then combine the per file tables using [concat_tables](https://arrow.apache.org/docs/python/generated/pyarrow.concat_tables.html).  I can set-up some benchmarks to comapre the two approaches, but it seems like from the answer on the user list that it's not currently possible if there is schema evolution across the files.

Reply:
I would be curious to see perf impact if we just made 10 copies of a single parquet file and read it in one go vs sequentially w/ a concat at the end.

Its too bad that the support for this use case in Arrow seems to be a future enhancement. Should be relatively easy to take it up once Arrow supports it.

Reply:
I just did some initial testing and I'm not seeing a significant difference between passing a list of files to Datasets vs. using a python threadpool per file and then concating the resulting tables. Let me come up with some more test cases and see if there are some where the arrow threading does better.  Using some internal data I read about ~90M rows in about 9 secs both ways. Test was run on a r4.16xlarge instance with EBS storage.  I'll try and setup something reproducible for you to try out

Reply:
Ok, thats good. I just wanted to get a broad idea to ensure we understand where the python reader will work/fail. So the bottleneck is really comign down to memory then as there is currently no way to page or dynamically load these per-file arrow tables

Reply:
Check this out and let me know what you think:
https://www.kaggle.com/tedgooch/notebook2b16380982

Reply:
thanks for the through investigation! Really interesting how close they are in a lot of cases.

Reply:
yeah just wondering if this is primarily IO bound, so the diff between C threads and python threads is negligible.  Maybe in cases where there is more cpu intense work going on this won't track so closely.   In any case, I think this is an ok design until we can get better native support for schema normalization in arrow itself.  What do you think?  Any other outstanding items before this is ready for merge?

Reply:
Yeah, I will do another quick pass now but I think its ready to go